src/Data/ExtractShardKey.ts

```diff
@@ -1,6 +1,44 @@
 import { config } from '../Config'
 import { AccountType } from '../shardeum/calculateAccountHash'
 import { InternalTXType } from '../shardeum/verifyGlobalTxReceipt'
+import * as NodeList from '../NodeList'
+import { getJson } from '../P2P'
+import * as Utils from '../Utils'
+
+export async function crackKeyFromSecureAccount(tx: any): Promise<string> {
+  // Define the query function for robustQuery
+  const queryFn = async (node: NodeList.ConsensusNodeInfo): Promise<any[]> => {
+    const response = (await getJson(`http://${node.ip}:${node.port}/secure_accounts`)) as any
+
+    if (response.accounts) {
+      return response.accounts
+    }
+
+    throw new Error(`Secure account data not found on node ${node.ip}`)
+  }
+
+  // Use robustQuery to fetch secure accounts
+  const activeNodes = NodeList.getActiveList()
+  const result = await Utils.robustQuery(activeNodes, queryFn)
+
+  if (!result.value || !Array.isArray(result.value)) {
+    throw new Error(`Unable to retrieve secure accounts`)
+  }
+
+  // Find the matching secure account from the result
+  const matchingAccount = result.value.find(
+    (account: any) => account.secureAccountConfig.Name === tx.accountName
+  )
+
+  if (!matchingAccount) {
+    throw new Error(`Secure account ${tx.accountName} not found`)
+  }
+
+  const secureAccountConfig = matchingAccount.secureAccountConfig
+
+  // Return the SourceFundsAddress formatted as a Shardus address
+  return toShardusAddress(secureAccountConfig.SourceFundsAddress, AccountType.Account)
```

Comment:
should the account type be `Account` or `SecureAccount`?

Reply:
```
export function crack(tx: InternalTx): CrackedData {
  if (!secureAccountDataMap.has(tx.accountName)) {
    throw new Error(`Secure account ${tx.accountName} not found`);
  }
  return {
    sourceKeys: [
      toShardusAddress(secureAccountDataMap.get(tx.accountName).SourceFundsAddress, AccountType.Account),
      toShardusAddress(secureAccountDataMap.get(tx.accountName).SecureAccountAddress, AccountType.SecureAccount),
    ],
    targetKeys: [
      toShardusAddress(secureAccountDataMap.get(tx.accountName).RecipientFundsAddress, AccountType.Account)
    ]
  }
}
```

The first key is of type account which is relevant to us